fcrepo/fcrepo
# Adds Stats endpoint with initial pass at

**Adds Stats endpoint with initial pass at  stats functionality**
* * *

**JIRA Ticket**:  https://fedora-repository.atlassian.net/browse/FCREPO-3801

* Other Relevant Links (Mailing list discussion, related pull requests, etc.)

# What does this Pull Request do?
Adds a /rest/fcr:stats endpoint  which returns total resource count, total binary count and a breakdown of the binary resource and byte count by mimetype.  
Use the mime_types query param to limit the binary count to specific mimetype.

# How should this be tested?
Start fedora
```
mvn clean jetty:run -pl fcrepo-webapp
```
Query the stats endpoint with an empty repo:
```
curl  -u fedoraAdmin:fedoraAdmin http://localhost:8080/rest/fcr:stats | jq
```
verify it returns this:
```
{
  "resource_count": 0,
  "binaries": {
    "mime_types": [],
    "byte_count": 0,
    "resource_count": 0
  },
  "all_resources": {
    "rdf_types": []
  }
}
```

Add a several resources of different types and in different quantities

```
for i in `seq 1 5`; do curl -u fedoraAdmin:fedoraAdmin http://localhost:8080/rest/hello-world-text-$i  -H "Content-Type: text/plain"  -X PUT --data "hello world text"; done; 

for i in `seq 1 7`; do curl -u fedoraAdmin:fedoraAdmin http://localhost:8080/rest/hello-world-audio-$i  -H "Content-Type: audio/mp4"  -X PUT --data "hello world audio"; done;

for i in `seq 1 10`; do curl -u fedoraAdmin:fedoraAdmin http://localhost:8080/rest/hello-world-container-$i  -X PUT; done;

```

Query the stats endpoint:
```
curl  -u fedoraAdmin:fedoraAdmin http://localhost:8080/rest/fcr:stats | jq
```
verify it returns this:
```
{
  "resource_count": 22,
  "binaries": {
    "mime_types": [
      {
        "resource_count": 7,
        "byte_count": 119,
        "mime_type": "audio/mp4"
      },
      {
        "resource_count": 5,
        "byte_count": 80,
        "mime_type": "text/plain"
      }
    ],
    "resource_count": 12,
    "byte_count": 199
  },
  "all_resources": {
    "rdf_types": [
      {
        "resource_count": 12,
        "byte_count": 199,
        "resource_type": "http://fedora.info/definitions/v4/repository#Binary"
      },
      {
        "resource_count": 10,
        "byte_count": 0,
        "resource_type": "http://fedora.info/definitions/v4/repository#Container"
      },
      {
        "resource_count": 22,
        "byte_count": 199,
        "resource_type": "http://fedora.info/definitions/v4/repository#Resource"
      },
      {
        "resource_count": 22,
        "byte_count": 199,
        "resource_type": "http://mementoweb.org/ns#OriginalResource"
      },
      {
        "resource_count": 22,
        "byte_count": 199,
        "resource_type": "http://mementoweb.org/ns#TimeGate"
      },
      {
        "resource_count": 10,
        "byte_count": 0,
        "resource_type": "http://www.w3.org/ns/ldp#BasicContainer"
      },
      {
        "resource_count": 10,
        "byte_count": 0,
        "resource_type": "http://www.w3.org/ns/ldp#Container"
      },
      {
        "resource_count": 12,
        "byte_count": 199,
        "resource_type": "http://www.w3.org/ns/ldp#NonRDFSource"
      },
      {
        "resource_count": 10,
        "byte_count": 0,
        "resource_type": "http://www.w3.org/ns/ldp#RDFSource"
      },
      {
        "resource_count": 22,
        "byte_count": 199,
        "resource_type": "http://www.w3.org/ns/ldp#Resource"
      }
    ]
  }
}
```
Return by mime type:  curl  -u fedoraAdmin:fedoraAdmin "http://localhost:8080/rest/fcr:stats/binaries" | jq


```
{
  "mime_types": [
    {
      "resource_count": 7,
      "byte_count": 119,
      "mime_type": "audio/mp4"
    },
    {
      "resource_count": 5,
      "byte_count": 80,
      "mime_type": "text/plain"
    }
  ],
  "resource_count": 12,
  "byte_count": 199
}

```

Return only audio/mp4  mime type:
```
curl  -u fedoraAdmin:fedoraAdmin "http://localhost:8080/rest/fcr:stats/binaries?mime_type=audio/mp4" | jq
```

Verify it returns this:

```
{
  "mime_types": [
    {
      "resource_count": 7,
      "byte_count": 119,
      "mime_type": "audio/mp4"
    }
  ],
  "resource_count": 7,
  "byte_count": 119
}

```

Return stats by rdf type only:
```
 curl  -u fedoraAdmin:fedoraAdmin "http://localhost:8080/rest/fcr:stats/rdf-types" | jq
```
verify it returns:
```
{
  "rdf_types": [
    {
      "resource_count": 12,
      "byte_count": 199,
      "resource_type": "http://fedora.info/definitions/v4/repository#Binary"
    },
    {
      "resource_count": 10,
      "byte_count": 0,
      "resource_type": "http://fedora.info/definitions/v4/repository#Container"
    },
    {
      "resource_count": 22,
      "byte_count": 199,
      "resource_type": "http://fedora.info/definitions/v4/repository#Resource"
    },
    {
      "resource_count": 22,
      "byte_count": 199,
      "resource_type": "http://mementoweb.org/ns#OriginalResource"
    },
    {
      "resource_count": 22,
      "byte_count": 199,
      "resource_type": "http://mementoweb.org/ns#TimeGate"
    },
    {
      "resource_count": 10,
      "byte_count": 0,
      "resource_type": "http://www.w3.org/ns/ldp#BasicContainer"
    },
    {
      "resource_count": 10,
      "byte_count": 0,
      "resource_type": "http://www.w3.org/ns/ldp#Container"
    },
    {
      "resource_count": 12,
      "byte_count": 199,
      "resource_type": "http://www.w3.org/ns/ldp#NonRDFSource"
    },
    {
      "resource_count": 10,
      "byte_count": 0,
      "resource_type": "http://www.w3.org/ns/ldp#RDFSource"
    },
    {
      "resource_count": 22,
      "byte_count": 199,
      "resource_type": "http://www.w3.org/ns/ldp#Resource"
    }
  ]
}
```



Return rdf types filtered by rdf type (ie show me rdf types for binaries only:
```
 curl  -u fedoraAdmin:fedoraAdmin "http://localhost:8080/rest/fcr:stats/rdf-types?rdf_type=http://www.w3.org/ns/ldp%23NonRDFSource" | jq
```
verify it returns:
```
{
  "rdf_types": [
    {
      "resource_count": 12,
      "byte_count": 199,
      "resource_type": "http://www.w3.org/ns/ldp#NonRDFSource"
    }
  ]
}
```

# Interested parties
Tag (@ mention) interested parties or, if unsure, @fcrepo/committers
